apiextensions/v1alpha1/types_compatibilityrequirement.go

```diff
@@ -0,0 +1,391 @@
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2479
+type CompatibilityRequirement struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitzero"`
+
+	// spec is the specification of the desired behavior of the Compatibility Requirement.
+	// +required
+	Spec CompatibilityRequirementSpec `json:"spec,omitzero"`
+
+	// status is the most recently observed status of the Compatibility Requirement.
+	// +optional
+	Status CompatibilityRequirementStatus `json:"status,omitzero"`
+}
+
+// CompatibilityRequirementSpec is the specification of the desired behavior of the Compatibility Requirement.
+type CompatibilityRequirementSpec struct {
+	// compatibilitySchema defines the schema used by
+	// customResourceDefinitionSchemaValidation and objectSchemaValidation.
+	// This field is required.
+	// +required
+	CompatibilitySchema CompatibilitySchema `json:"compatibilitySchema,omitempty,omitzero"`
+
+	// customResourceDefinitionSchemaValidation ensures that updates to the
+	// installed CRD are compatible with this compatibility requirement. If not
+	// specified, admission of the target CRD will not be validated.
+	// This field is optional.
+	// +optional
+	CustomResourceDefinitionSchemaValidation CustomResourceDefinitionSchemaValidation `json:"customResourceDefinitionSchemaValidation,omitempty,omitzero"`
+
+	// objectSchemaValidation ensures that matching resources conform to
+	// compatibilitySchema. If not specified, admission of matching resources
+	// will not be validated.
+	// This field is optional.
+	// +optional
+	ObjectSchemaValidation ObjectSchemaValidation `json:"objectSchemaValidation,omitempty,omitzero"`
+}
+
+// CRDDataType indicates the type of the CRD data.
+type CRDDataType string
+
+const (
+	// CRDDataTypeYAML indicates that the CRD data is in YAML format.
+	CRDDataTypeYAML CRDDataType = "YAML"
+)
+
+// CRDData contains the complete definition of a CRD.
+type CRDData struct {
+	// type indicates the type of the CRD data. The only supported type is YAML.
+	// This field is required.
+	// +kubebuilder:validation:Enum=YAML
+	// +required
+	Type CRDDataType `json:"type,omitempty"`
+
+	// data contains the complete definition of the CRD.
+	// This field is required.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1572864
+	// +required
+	Data string `json:"data,omitempty"`
+}
+
+// APIVersionSetType specifies a method for selecting the set of API versions to require.
+// +kubebuilder:validation:Enum=StorageOnly;All
+type APIVersionSetType string
+
+const (
+	APIVersionSetTypeStorageOnly APIVersionSetType = "StorageOnly"
+	APIVersionSetTypeAll         APIVersionSetType = "All"
+)
+
+// APIVersionString is a string representing a kubernetes API version.
```

Comment:
Are API versions regexable? Does the CRD storage check them against some pattern?

Reply:
I think I briefly flirted with this. However, these are editable. If the caller gets it wrong it can be corrected without any upgrade impact when it becomes apparent that they entered the wrong version. The same isn't true when somebody calls their new version `v1beta2alpha1` and we didn't think of that in advance. IOW the impact of over-validation is higher than the impact of under-validation. I'm inclined to leave it alone unless there's an existing well-trodden path.

Reply:
> I bet there's already a CEL function for it, though, so it's easy enough to add.

Correct, there is, did you work this out or do you need a pointer?